### PR TITLE
Distributed training example

### PR DESCRIPTION
**Pull request by IsaacWarren**
## Changes included in this PR
Training example with some changes to the torch train API to make distributed GPU training work.

## Testing strategy
Locally and on multi-gpu instance

## User facing changes

In docs
## Checklist
- [ ] Pipelines passed before requesting review. To run CI you must include `[run CI]` in your commit message.
- [x] I am familiar with the [Contributing Guide](https://github.com/bodo-ai/Bodo/blob/main/CONTRIBUTING.md) 
- [x] I have installed + ran pre-commit hooks.

### CODE REVIEW

**Review by DrTodd13 — APPROVED** [review]
LGTM.  thanks Isaac.

File: examples/ML/pandas_ddp_example.py [hunk 1]
```diff
@@ -0,0 +1,298 @@
...
+    train_dataset = PandasDataset(train_df, device)
+    val_dataset = PandasDataset(val_df, device)
+    test_dataset = PandasDataset(test_df, device)
+    train_sampler = BodoDistributedSampler(
+        train_dataset,
+        worker_ranks=gpu_ranks
+        if accelerators_used
+        else list(range(MPI.COMM_WORLD.Get_size())),
+    )
+    val_sampler = BodoDistributedSampler(
```
Comment (scott-routledge2) [inline examples/ML/pandas_ddp_example.py:244]: Minor but the val/test sampler don't need to shuffle.

File: examples/ML/pandas_ddp_example.py [hunk 2]
```diff
@@ -0,0 +1,298 @@
...
+import torch.distributed as dist
+import torch.distributed.checkpoint
+import tqdm
+from torch import nn
+from torch.optim import Adam
+from torch.utils.data import DataLoader
+from transformers import BertModel, BertTokenizer
+
+LR = 1e-6
+EPOCHS = 1
```
Comment (scott-routledge2) [inline examples/ML/pandas_ddp_example.py]: We probably are going to want to show multiple epochs for the actual example

File: examples/ML/pandas_ddp_example.py [hunk 3]
```diff
@@ -0,0 +1,298 @@
...
+NUM_CLASSES = 5
+SEQ_LENGTH = 512
+BATCH_SIZE = 25
```
Comment (scott-routledge2) [inline examples/ML/pandas_ddp_example.py]: Why BATCH_SIZE=25?

File: examples/ML/pandas_ddp_example.py [hunk 4]
```diff
@@ -0,0 +1,298 @@
...
+    validation(model, test_loader, loss_fn)
+
+
+
+if __name__ == "__main__":
+    tokenizer = BertTokenizer.from_pretrained('bert-base-uncased')
+    train_df, val_df, test_df = prepare_datasets(tokenizer)
+    train_df.execute_plan()
+    val_df.execute_plan()
+    test_df.execute_plan()
```
Comment (scott-routledge2) [inline examples/ML/pandas_ddp_example.py]: Also was wondering why we need to execute plans first

File: examples/ML/pandas_ddp_example.py (hunk 3 above)
Comment (IsaacWarren) [reply examples/ML/pandas_ddp_example.py]: Discussed offline

File: examples/ML/pandas_ddp_example.py (hunk 4 above)
Comment (IsaacWarren) [reply examples/ML/pandas_ddp_example.py]: There was some weird error sending them to workers if I didn't
